gtsam/inference/BayesTree-inst.h

```diff
@@ -63,20 +63,40 @@ namespace gtsam {
   }
 
   /* ************************************************************************* */
-  template<class CLIQUE>
-  void BayesTree<CLIQUE>::saveGraph(const std::string &s, const KeyFormatter& keyFormatter) const {
-    if (roots_.empty()) throw std::invalid_argument("the root of Bayes tree has not been initialized!");
-    std::ofstream of(s.c_str());
-    of<< "digraph G{\n";
-    for(const sharedClique& root: roots_)
-      saveGraph(of, root, keyFormatter);
-    of<<"}";
+  template <class CLIQUE>
+  void BayesTree<CLIQUE>::dot(std::ostream& os,
+                              const KeyFormatter& keyFormatter) const {
+    if (roots_.empty())
+      throw std::invalid_argument(
+          "the root of Bayes tree has not been initialized!");
+    os << "digraph G{\n";
+    for (const sharedClique& root : roots_) dot(os, root, keyFormatter);
+    os << "}";
+    std::flush(os);
+  }
+
+  /* ************************************************************************* */
+  template <class CLIQUE>
+  std::string BayesTree<CLIQUE>::dot(const KeyFormatter& keyFormatter) const {
```

Comment:
Seeing this repeat in multiple places, do you think it is worth having this as a trait? The `Testable` trait has a requirement for `print` so we could have a similar requirement for `dot(ostream&, KeyFormatter&)`.
Just a thought.

Reply:
That's a good idea. I'll add a TODO and issue.

Reply:
Actually, I started on the issue but I don't see the reason, after all. traits are useful if they are used for some functionality. We can make traits if it serves a purpose, later.